src/toolkits/object_detection/one_shot_object_detection/one_shot_object_detector.cpp

```diff
@@ -63,181 +62,200 @@ flex_dict build_annotation( ParameterSampler &parameter_sampler,
+    size_t segment_end = ((segment_id + 1) * backgrounds.size()) / nsegments;
+    size_t row_number = segment_start;
+    for (const auto &background_ft :
+         backgrounds.range_iterator(segment_start, segment_end)) {
       row_number++;
-      flex_image flex_background = background_ft.to<flex_image>();
-      if (!(flex_background.is_decoded())) {
-        decode_image_inplace(flex_background);
-      }
-      size_t background_width = flex_background.m_width;
-      size_t background_height = flex_background.m_height;
-      size_t background_channels = flex_background.m_channels;
-
-      ParameterSampler parameter_sampler = ParameterSampler(
-                                              background_width,
-                                              background_height,
-                                              (background_width-object.m_width)/2,
-                                              (background_height-object.m_height)/2);
-
-      flex_dict annotation = build_annotation(parameter_sampler, label,
-                                              object.m_width, object.m_height,
-                                              seed+row_number);
-
-      DASSERT_TRUE(flex_background.get_image_data() != nullptr);
-      DASSERT_TRUE(object.is_decoded());
-      DASSERT_TRUE(flex_background.is_decoded());
-
-      boost::gil::rgba8_image_t::view_t starter_image_view = create_starter_image_view(object);
-
-      boost::gil::rgb8_image_t::view_t background_view = interleaved_view(
-        background_width,
-        background_height,
-        (boost::gil::rgb8_pixel_t*)(flex_background.get_image_data()),
-        background_channels * background_width // row length in bytes
-        );
-
-      images.push_back(
-        create_synthetic_image(starter_image_view, background_view, parameter_sampler, object)
-      );
-      annotations.push_back(annotation);
-
-      if (verbose) {
-        std::ostringstream d;
-        // For pretty printing, floor percent done
-        // resolution to the nearest .25% interval.  Do this by multiplying by
-        // 400, then do integer division by the total size, then float divide
-        // by 4.0
-        int augmented_rows_completed = (input_row_index * backgrounds_size) + row_number;
-        d << augmented_rows_completed * 400 / total_augmented_rows / 4.0 << '%';
-        table.print_progress_row(augmented_rows_completed,
-                                  augmented_rows_completed,
-                                  progress_time(), d.str());
+      flex_image flex_background =
+          image_util::decode_image(background_ft.to<flex_image>());
+      for (const auto &row : decompressed_data.range_iterator()) {
+        // go through all the starter images and create augmented images for
+        // all starter images and the respective chunk of background images
+        flex_image object = row[image_column_index].to<flex_image>();
```

Comment:
If we are confident that this is an image column, then we should use `const flex_image& object = row[image_column_index].get<flex_image>();`.

Reply:
Fixed.